dragonfly/actions/keyboard.py

```diff
@@ -34,68 +31,94 @@
                                          send_input_array)
 
 
-#---------------------------------------------------------------------------
-# Typeable class.
-
 class Typeable(object):
+    """Container for keypress events."""
 
-    __slots__ = ("_code", "_modifiers", "_name")
+    __slots__ = ("_code", "_modifiers", "_name", "_is_text")
 
-    def __init__(self, code, modifiers=(), name=None):
+    def __init__(self, code, modifiers=(), name=None, is_text=False):
+        """Set keypress information."""
         self._code = code
         self._modifiers = modifiers
         self._name = name
+        self._is_text = is_text
 
     def __str__(self):
-        return "%s(%s)" % (self.__class__.__name__, self._name) + repr(self.events())
+        """Return information useful for debugging."""
+        return ("%s(%s)" % (self.__class__.__name__, self._name) +
+                repr(self.events()))
 
     def on_events(self, timeout=0):
-        events = [(m, True, 0) for m in self._modifiers]
-        events.append((self._code, True, timeout))
+        """Return events for pressing this key down."""
+        if self._is_text:
+            events = [(self._code, True, timeout, True)]
+        else:
+            events = [(m, True, 0) for m in self._modifiers]
```

Comment:
Did you observe that modifiers no longer work when using unicode, or was this just a separate cleanup to match the documented API? In the long-term, at least, it seems helpful to decouple these changes, so that control keys behave the same across apps with and without support for unicode keystrokes, in case people come to depend on the new behavior. In other words, we can either remove the modifier key stuff entirely, or make it a separate option for backwards-compatibility.

Also, if the modifier stuff doesn't work but also doesn't break anything, it seems like we could keep it in to avoid adding so many branches throughout the code.

Reply:
It was an observation I made. I think I understand why. It helps to look at the way Windows is processing this in terms of message loop events and with the understanding that Windows represents characters as UTF-16LE internally and does conversion depending on what the window is expecting. Normally valid virtual keycodes are in the range from 1 to 254, inclusive.

I'll use typing the lowercase letter `L` as an example.
 * Using normal keyboard emulation:
   1. `SendInput` is called with the virtual keycode value `76`, the appropriate scancode returned by `MapVirtualKey`, and a `flags` value of `8`(indicating a hardware scancode is present).
   2. Windows uses this information to send the window a keydown event message with the virtual keycode value `76` (`VK_L`) and no control modifiers.
   3. This is followed by a character message with the appropriate character code for `L`.
   4. The call to `SendInput` is repeated with the value `2` OR'd with the `flags` (indicating the key is being released).
   5. The sequence ends with the window receiving a keyup event message with the virtual keycode value `76` and no control modifiers.

* Using Unicode keyboard emulation:
   1. `SendInput` is called with the virtual keycode value `0`, the scancode set to the UTF-16LE value of the character, and a `flags` value of `4`(indicating the scancode should be set to the requested character).
   2. Windows uses this information to send the window a keydown event message with the virtual keycode value `231` (`VK_PACKET`) and no control modifiers.
   3. This is followed by a character message with the appropriate character code for `L`.
   4. The call to `SendInput` is repeated with the value `2` OR'd with the `flags` (indicating the key is being released).
   5. The sequence ends with the window receiving a keyup event message with the virtual keycode value `231` and no control modifiers.


In the normal case, the window receives the correct virtual keycode for the character typed. In the Unicode case, however, the virtual keycode sent to the window will always be `231` (`VK_PACKET`) regardless of which character is actually sent. For most applications in most cases the character message is the part the program actually uses to understand what is typed. Things like keyboard accelerators tend to be mapped to a combination of modifiers with specific virtual keycodes, which is where the problem lies.

Reply:
This makes a lot of sense, thank you for the detailed explanation.